src/NHibernate/Linq/NhLinqExpression.cs

```diff
@@ -88,13 +88,16 @@ public IASTNode Translate(ISessionFactoryImplementor sessionFactory, bool filter
 
 			ParameterDescriptors = requiredHqlParameters.AsReadOnly();
 
-			CanCachePlan = CanCachePlan &&
-				// If some constants do not have matching HQL parameters, their values from first query will
-				// be embedded in the plan and reused for subsequent queries: do not cache the plan.
-				!ParameterValuesByName
+			if (QueryMode == QueryMode.Select && CanCachePlan)
+			{
+				CanCachePlan =
+					// If some constants do not have matching HQL parameters, their values from first query will
+					// be embedded in the plan and reused for subsequent queries: do not cache the plan.
+					!ParameterValuesByName
 					.Keys
 					.Except(requiredHqlParameters.Select(p => p.Name))
 					.Any();
+			}
```

Comment:
Shouldn't CanCachePlan become false here?

Reply:
When? Fix simply ignores logic that was introduced #1544 and DML queries are **always** cached.

My understanding that "non-parameterized constants issue" is not relevant for DML queries as no result transformers with embedded constants are possible for DML.

Reply:
It is relevant. The original tests do not cover all cases of the issue and just focusing on the constants in select statement.

Reply:
Can you provide an example?

Reply:
If you are sure  that DML queries can be affected you can close this PR. But currently DML query plans are never cached as `requiredHqlParameters` contain only parameters for Select part (seems [intentional](https://github.com/nhibernate/nhibernate-core/blob/a41010f0611e93fc13c56af615b86b4bb90ad01d/src/NHibernate/Linq/DmlExpressionRewriter.cs#L103-L105) behavior)

Reply:
I tried to come up with a query, but I could not.

Reply:
Cached linq has better than non cached, I don't figure out why but its quite slow than hql.

https://github.com/nhibernate/nhibernate-core/issues/2222#issuecomment-534267789